common/src/main/java/org/conscrypt/AbstractConscryptSocket.java

```diff
@@ -588,10 +588,11 @@ final PeerInfoProvider peerInfoProvider() {
     abstract void setApplicationProtocolSelector(ApplicationProtocolSelectorAdapter selector);
 
     final void notifyHandshakeCompletedListeners() {
-        if (!listeners.isEmpty()) {
+        List<HandshakeCompletedListener> listenersCopy = new ArrayList<>(listeners);
```

Comment:
This will throw NPE if it turns out to be `new ArrayList<>(null)`. Perhaps this:

```java
List<HandshakeCompletedListener> listenersLocal = listeners;
if (listenersLocal != null) {
  List<HandshakeCompletedListener> listenersCopy = new ArrayList<>(listenersLocal);
  ...
```

That would guard against the field being set to `null` between check and use as well as the listener being removed in the middle of iteration. The `.isEmpty()` could be kept as a small optimization.

Reply:
Note that listeners is final and has an initialiser, so if it's null then the app has bigger problems :)

Reply:
Great. Didn't check that :)